crates/sui-graphql-rpc/src/raw_query.rs

```diff
@@ -0,0 +1,199 @@
+///    the need to specify types explicitly, as is typically required with Diesel's
+///    `sql_query.bind`.
+///
+/// 2. **Query String Merging**: Can be used to represent and merge query strings and their
+///    associated bindings. Placeholder strings and bindings are applied in sequential order.
+///
+/// Note: `RawQuery` only supports binding string values, as interpolating raw strings directly
+/// increases exposure to SQL injection attacks.
+#[derive(Clone)]
+pub(crate) struct RawQuery {
+    /// The `SELECT` clause of the query.
+    select: String,
+    /// The `WHERE` clause of the query.
+    where_: Option<String>,
+    /// The `ORDER BY` clause of the query.
+    order_by: Vec<String>,
+    /// The `GROUP BY` clause of the query.
+    group_by: Vec<String>,
+    /// The `LIMIT` clause of the query.
+    limit: Option<i64>,
+    /// The list of string binds for this query.
+    binds: Vec<String>,
+}
+
+impl RawQuery {
+    /// Constructs a new `RawQuery` with the given `SELECT` clause and binds.
+    pub(crate) fn new(select: impl Into<String>, binds: Vec<String>) -> Self {
+        Self {
+            select: select.into(),
+            where_: None,
+            order_by: Vec::new(),
+            group_by: Vec::new(),
+            limit: None,
+            binds,
+        }
+    }
+
+    /// Adds a `WHERE` condition to the query, combining it with existing conditions using `AND`.
+    pub(crate) fn filter<T: std::fmt::Display>(mut self, condition: T) -> Self {
+        self.where_ = match self.where_ {
+            Some(where_) => Some(format!("({}) AND {}", where_, condition)),
+            None => Some(condition.to_string()),
+        };
+
+        self
+    }
+
+    /// Adds a `WHERE` condition to the query, combining it with existing conditions using `OR`.
+    #[allow(dead_code)]
+    pub(crate) fn or_filter<T: std::fmt::Display>(mut self, condition: T) -> Self {
+        self.where_ = match self.where_ {
+            Some(where_) => Some(format!("({}) OR {}", where_, condition)),
+            None => Some(condition.to_string()),
+        };
+
+        self
+    }
+
+    /// Adds an `ORDER BY` clause to the query.
+    pub(crate) fn order_by<T: ToString>(mut self, order: T) -> Self {
+        self.order_by.push(order.to_string());
+        self
+    }
+
+    /// Adds a `GROUP BY` clause to the query.
+    pub(crate) fn group_by<T: ToString>(mut self, group: T) -> Self {
+        self.group_by.push(group.to_string());
+        self
+    }
+
+    /// Adds a `LIMIT` clause to the query.
+    pub(crate) fn limit(mut self, limit: i64) -> Self {
+        self.limit = Some(limit);
+        self
+    }
+
+    /// Adds the `String` value to the list of binds for this query.
+    pub(crate) fn bind_value(&mut self, condition: String) {
+        self.binds.push(condition);
+    }
+
+    /// Constructs the query string and returns it along with the list of binds for this query.
```

Comment:
Worth calling out that this isn't intended for people to call under normal operation, but is exposed for the `query!` macro to use?